pinot-common/src/main/java/com/linkedin/pinot/common/segment/fetcher/HttpsSegmentFetcher.java

```diff
@@ -0,0 +1,169 @@
+ *
+ * At this time, only PKCS12 files are supported for client certificates, and it is assumed that the
+ * server is presenting an X509 certificate.
+ */
+public class HttpsSegmentFetcher implements SegmentFetcher {
+  SSLSocketFactory _sslSocketFactory;
+
+  private static final String SECURITY_ALGORITHM = "TLS";
+  private static final String CERTIFICATE_TYPE = "X509";
+  private static final String KEYSTORE_TYPE = "PKCS12";
+  private static final String KEYMANAGER_FACTORY_ALGORITHM = "SunX509";
+
+  private static String CONFIG_OF_SERVER_CA_CERT = "ssl.server.ca-cert";
+  private static String CONFIG_OF_CLIENT_PKCS12_FILE = "ssl.client.pkcs12.file";
+  private static String CONFIG_OF_CLIENT_PKCS12_PASSWORD = "ssl.client.pkcs12.password";
+
+  public static Logger LOGGER = LoggerFactory.getLogger(HttpsSegmentFetcher.class);
+
+  @Override
+  public void init(Configuration configs) {
+    try {
+      TrustManager[] trustManagers = setupTrustManagers(configs);
+      KeyManager[] keyManagers = setupKeyManagers(configs);
+
+      SSLContext sslContext = SSLContext.getInstance(SECURITY_ALGORITHM);
+      sslContext.init(keyManagers, trustManagers, null);
+      _sslSocketFactory = sslContext.getSocketFactory();
+    } catch (Exception e) {
+      Utils.rethrowException(e);
+    }
+  }
+
+  private TrustManager[] setupTrustManagers(Configuration configs)
+      throws CertificateException, KeyStoreException, IOException, NoSuchAlgorithmException {
+    // This is the cert authority that validates server's cert, so we need to put it in our
+    // trustStore.
+    String serverCACertFile = configs.getString(CONFIG_OF_SERVER_CA_CERT);
+    if (serverCACertFile == null) {
+      LOGGER.warn("Https server CA Certificate file not confugured (" + CONFIG_OF_SERVER_CA_CERT + "). All servers will be trusted!");
+      TrustManager[] trustAllCerts = new TrustManager[] {
+          new X509TrustManager() {
+            @Override
+            public void checkClientTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {
+            }
+
+            @Override
+            public void checkServerTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {
+            }
+
+            @Override
+            public X509Certificate[] getAcceptedIssuers() {
+              return new X509Certificate[0];
+            }
+          }
+      };
+      return trustAllCerts;
+    }
+    FileInputStream is = new FileInputStream(new File(serverCACertFile));
+    CertificateFactory certificateFactory = CertificateFactory.getInstance(CERTIFICATE_TYPE);
+    X509Certificate cert = (X509Certificate) certificateFactory.generateCertificate(is);
+
+    String serverKey = "https-server";
+    KeyStore trustStore = KeyStore.getInstance(KeyStore.getDefaultType());
+    trustStore.load(null);
+    trustStore.setCertificateEntry(serverKey, cert);
+
+    TrustManagerFactory tmf = TrustManagerFactory.getInstance(CERTIFICATE_TYPE);
```

Comment:
Could we leave some comments regarding this logic? Spoke offline and it made sense, but might be difficult in the future to decrypt

Reply:
I have added some links. Most SSL APIs don't even have visible names for the method arguments, so there is little that can be done to name the strings used. I have tried my best after reading some of the links that I have added in the class.